mne/preprocessing/_csd.py

```diff
@@ -153,9 +151,18 @@ def compute_current_source_density(inst, sphere='auto', lambda2=1e-5,
         raise ValueError('Zero or infinite position found in chs')
     pos -= (x, y, z)
 
-    G = _calc_g(np.dot(pos, pos.T), stiffness=stiffness,
+    # Original code does not project onto the sphere, probably not good:
+    # pos /= radius
+    # from scipy.spatial.distance import squareform, pdist
+    # cos_dist = 1 - squareform(pdist(pos, 'sqeuclidean')) / 2.
+    # Project onto the sphere and do the distance (effectively)
+    pos /= np.linalg.norm(pos, axis=1, keepdims=True)
+    cos_dist = np.clip(np.dot(pos, pos.T), -1, 1)
+    del pos
+
+    G = _calc_g(cos_dist, stiffness=stiffness,
                 n_legendre_terms=n_legendre_terms)
-    H = _calc_h(np.dot(pos, pos.T), stiffness=stiffness,
+    H = _calc_h(cos_dist, stiffness=stiffness,
```

Comment:
... *and* @alexrockhill our old code did use the actual cosine distance, so that's both the name and someone else's implementation that make it seem like it should be the cosine similarity, not one minus the squared euclidean distance divided by two. But this is just anecdotal, someone might need to go to the papers to check.